File: packages/typescript/src/parser.ts
import { AstNode, BlockNode, DocumentNode } from "./ast.js";
import { ParserException } from "./error.js";
import { Lexer } from "./lexer.js";
import { Token, TokenType } from "./token.js";

export class Parser {
  private current: Token;

  constructor(private readonly lexer: Lexer) {
    this.current = lexer.nextToken();
  }

  parse(): DocumentNode {
    const blocks: BlockNode[] = [];
    while (this.current.type !== TokenType.EOF) blocks.push(this.parseBlock());
    return { kind: "document", blocks };
  }

  private parseBlock(): BlockNode {
    const name = this.current.value;
    this.eat(TokenType.Identifier);

    let argument: string | undefined;
    if (this.current.type === TokenType.String) {
      argument = this.current.value;
      this.eat(TokenType.String);
    }

    this.eat(TokenType.Do);

    const properties: Record<string, AstNode> = {};
    const blocks: Record<string, BlockNode> = {};
    const namedBlocks: BlockNode[] = [];

    while (this.current.type !== TokenType.End) {
      if (this.current.type === TokenType.EOF) throw this.err("missing 'end' for block");
      if (this.current.type !== TokenType.Identifier) throw this.err(`expected identifier, got ${this.current.type}`);

      const next = this.peekToken();
      if (next.type === TokenType.Do || next.type === TokenType.String) {
        const child = this.parseBlock();
        if (child.argument) {
          namedBlocks.push(child);
          blocks[`${child.name}:${child.argument}`] = child;
        } else blocks[child.name] = child;
      } else if (next.type === TokenType.Equal || next.type === TokenType.Dot) {
        const key = this.parsePropertyKey();
        this.eat(TokenType.Equal);
        properties[key] = this.parseValue();
      } else throw this.err(`invalid statement after '${this.current.value}'`);
    }

    this.eat(TokenType.End);
    const block: BlockNode = { kind: "block", name, properties, blocks };
    if (argument !== undefined) block.argument = argument;
    if (namedBlocks.length > 0) block.named_blocks = namedBlocks;
    return block;
  }

  private parsePropertyKey(): string {
    let key = this.current.value;
    this.eat(TokenType.Identifier);
    while (this.current.type === TokenType.Dot) {
      this.eat(TokenType.Dot);
      if (this.current.type !== TokenType.Identifier) throw this.err("expected identifier after dot");
      key += `.${this.current.value}`;
      this.eat(TokenType.Identifier);
    }
    return key;
  }

  private parseValue(): AstNode {
    if (this.current.type === TokenType.String) {
      const v = this.current.value; this.eat(TokenType.String); return { kind: "string", value: v };
    }
    if (this.current.type === TokenType.Number) {
      const v = Number(this.current.value); this.eat(TokenType.Number); return { kind: "number", value: v };
    }
    if (this.current.type === TokenType.Identifier) {
      const v = this.current.value; this.eat(TokenType.Identifier);
      if (v === "true") return { kind: "boolean", value: true };
      if (v === "false") return { kind: "boolean", value: false };
      return { kind: "string", value: v };
    }
    if (this.current.type === TokenType.LBracket) return this.parseArray();
    throw this.err(`unexpected token ${this.current.type}`);
  }

  private parseArray(): AstNode {
    this.eat(TokenType.LBracket);
    const elements: AstNode[] = [];
    if (this.current.type !== TokenType.RBracket) {
      elements.push(this.parseValue());
      while (this.current.type === TokenType.Comma) {
        this.eat(TokenType.Comma);
        elements.push(this.parseValue());
      }
    }
    this.eat(TokenType.RBracket);
    return { kind: "array", elements };
  }

  private eat(type: TokenType): void {
    if (this.current.type !== type) throw this.err(`expected ${type}, got ${this.current.type}`);
    this.current = this.lexer.nextToken();
  }

  private peekToken(): Token {
    const st = this.lexer.snapshot();
    const tok = this.lexer.nextToken();
    this.lexer.restore(st);
    return tok;
  }

  private err(message: string): ParserException {
    return new ParserException(message, this.current.line, this.current.column);
  }
}

File: packages/rust/src/parser.rs
use std::collections::BTreeMap;

use crate::ast::{AstNode, BlockNode, DocumentNode};
use crate::error::ParseError;
use crate::lexer::{Lexer, Token, TokenType};

pub struct Parser { lex: Lexer, cur: Token }

impl Parser {
    pub fn new(text: &str) -> Result<Self, ParseError> {
        let mut lex = Lexer::new(text);
        let cur = lex.next_token()?;
        Ok(Self { lex, cur })
    }

    pub fn parse(&mut self) -> Result<DocumentNode, ParseError> {
        let mut blocks = Vec::new();
        while self.cur.typ != TokenType::Eof { blocks.push(self.parse_block()?); }
        Ok(DocumentNode { kind: "document", blocks })
    }

    fn parse_block(&mut self) -> Result<BlockNode, ParseError> {
        let name = self.cur.value.clone(); self.eat(TokenType::Identifier)?;
        let argument = if self.cur.typ == TokenType::String { let v = self.cur.value.clone(); self.eat(TokenType::String)?; Some(v) } else { None };
        self.eat(TokenType::Do)?;

        let mut properties = BTreeMap::new();
        let mut blocks = BTreeMap::new();
        let mut named_blocks = Vec::new();
        let mut seen = Vec::<String>::new();

        while self.cur.typ != TokenType::End {
            if self.cur.typ == TokenType::Eof { return Err(self.err("missing end")); }
            if self.cur.typ != TokenType::Identifier { return Err(self.err("expected identifier")); }
            let next = self.peek_token()?;
            if next.typ == TokenType::Do || next.typ == TokenType::String {
                let child = self.parse_block()?;
                if child.argument.is_some() {
                    named_blocks.push(child.clone());
                } else { blocks.insert(child.name.clone(), child); }
            } else if next.typ == TokenType::Equal || next.typ == TokenType::Dot {
                let key = self.parse_key()?;
                self.ensure_key_valid(&key, &seen)?;
                seen.push(key.clone());
                self.eat(TokenType::Equal)?;
                properties.insert(key, self.parse_value()?);
            } else { return Err(self.err("invalid statement")); }
        }
        self.eat(TokenType::End)?;
        Ok(BlockNode { kind: "block", name, argument, properties, blocks, named_blocks })
    }

    fn parse_key(&mut self) -> Result<String, ParseError> {
        let mut key = self.cur.value.clone(); self.eat(TokenType::Identifier)?;
        while self.cur.typ == TokenType::Dot {
            self.eat(TokenType::Dot)?;
            if self.cur.typ != TokenType::Identifier { return Err(self.err("expected identifier after dot")); }
            key.push('.'); key.push_str(&self.cur.value); self.eat(TokenType::Identifier)?;
        }
        Ok(key)
    }

    fn parse_value(&mut self) -> Result<AstNode, ParseError> {
        match self.cur.typ {
            TokenType::String => { let v = self.cur.value.clone(); self.eat(TokenType::String)?; Ok(AstNode::String { kind: "string", value: v }) }
            TokenType::Number => {
                let v: f64 = self.cur.value.parse().map_err(|_| self.err("invalid number"))?;
                self.eat(TokenType::Number)?;
                Ok(AstNode::Number { kind: "number", value: v })
            }
            TokenType::Identifier => {
                let v = self.cur.value.clone(); self.eat(TokenType::Identifier)?;
                if v == "true" { Ok(AstNode::Boolean { kind: "boolean", value: true }) }
                else if v == "false" { Ok(AstNode::Boolean { kind: "boolean", value: false }) }
                else { Err(self.err("invalid bare value")) }
            }
            TokenType::LBracket => self.parse_array(),
            _ => Err(self.err("unexpected value")),
        }
    }

    fn parse_array(&mut self) -> Result<AstNode, ParseError> {
        self.eat(TokenType::LBracket)?;
        let mut elements = Vec::new();
        if self.cur.typ != TokenType::RBracket {
            elements.push(self.parse_value()?);
            while self.cur.typ == TokenType::Comma {
                self.eat(TokenType::Comma)?;
                if self.cur.typ == TokenType::RBracket { return Err(self.err("trailing comma in array")); }
                elements.push(self.parse_value()?);
            }
        }
        self.eat(TokenType::RBracket)?;
        Ok(AstNode::Array { kind: "array", elements })
    }

    fn eat(&mut self, typ: TokenType) -> Result<(), ParseError> {
        if self.cur.typ != typ { return Err(self.err("unexpected token")); }
        self.cur = self.lex.next_token()?;
        Ok(())
    }

    fn peek_token(&mut self) -> Result<Token, ParseError> {
        let st = self.lex.snapshot();
        let tok = self.lex.next_token()?;
        self.lex.restore(st);
        Ok(tok)
    }

    fn err(&self, m: &str) -> ParseError { ParseError::new(m, self.cur.line, self.cur.col) }

    fn ensure_key_valid(&self, key: &str, seen: &[String]) -> Result<(), ParseError> {
        if seen.iter().any(|k| k == key) { return Err(self.err("duplicate key")); }
        let parts: Vec<&str> = key.split('.').collect();
        for existing in seen {
            let ex: Vec<&str> = existing.split('.').collect();
            if is_prefix(&parts, &ex) || is_prefix(&ex, &parts) { return Err(self.err("key conflict")); }
        }
        Ok(())
    }
}

fn is_prefix(left: &[&str], right: &[&str]) -> bool {
    if left.len() >= right.len() { return false; }
    left.iter().enumerate().all(|(i, v)| right[i] == *v)
}
